cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java

```diff
@@ -146,57 +146,72 @@ protected MetricsRecord computeNext() {
-              .expireAfterAccess(CACHE_EXPIRE_MINUTES, TimeUnit.MINUTES)
-              .build(new CacheLoader<String, EmitterKey>() {
-                @Override
-                public EmitterKey load(String metric) throws Exception {
-                  return new EmitterKey(collectorKey, metric);
-                }
-              });
-
-          @Override
-          public void increment(String metricName, int value, String... tags) {
-            emitters.getUnchecked(keys.getUnchecked(metricName)).increment(value, tags);
-          }
-
-          @Override
-          public void gauge(String metricName, long value, String... tags) {
-            emitters.getUnchecked(keys.getUnchecked(metricName)).gauge(value, tags);
-          }
-        };
+        return new MetricsCollectorImpl(collectorKey);
       }
     };
   }
 
+  private final class MetricsCollectorImpl implements MetricsCollector {
+    // Cache for minimizing creating new MetricKey object.
+    private final LoadingCache<String, EmitterKey> keys;
+
+    private final CollectorKey collectorKey;
+
+    private MetricsCollectorImpl(final CollectorKey collectorKey) {
+      this.collectorKey = collectorKey;
+      this.keys = CacheBuilder.newBuilder()
+          .expireAfterAccess(CACHE_EXPIRE_MINUTES, TimeUnit.MINUTES)
+          .build(new CacheLoader<String, EmitterKey>() {
+            @Override
+            public EmitterKey load(String metric) throws Exception {
+              return new EmitterKey(collectorKey, metric);
+            }
+          });
+    }
+
+    @Override
+    public void increment(String metricName, int value) {
+      emitters.getUnchecked(keys.getUnchecked(metricName)).increment(value);
+    }
+
+    @Override
+    public void gauge(String metricName, long value) {
+      emitters.getUnchecked(keys.getUnchecked(metricName)).gauge(value);
+    }
+
+    @Override
+    public MetricsCollector childCollector(String tagName, String tagValue) {
+      ImmutableMap<String, String> tags = ImmutableMap.<String, String>builder()
+        .putAll(collectorKey.tags).put(tagName, tagValue).build();
+      return new MetricsCollectorImpl(new CollectorKey(collectorKey.scope, tags));
+    }
+
+    @Override
+    public MetricsCollector childCollector(Map<String, String> tags) {
+      ImmutableMap<String, String> allTags = ImmutableMap.<String, String>builder()
+        .putAll(collectorKey.tags).putAll(tags).build();
+      return new MetricsCollectorImpl(new CollectorKey(collectorKey.scope, allTags));
+    }
+  }
+
   /**
    * Inner class for cache key for looking up {@link MetricsCollector}.
    */
   private static final class CollectorKey {
     private final MetricsScope scope;
-    private final String context;
-    private final String runId;
+    private final Map<String, String> tags;
 
-    private CollectorKey(MetricsScope scope, String context, String runId) {
+    private CollectorKey(MetricsScope scope, Map<String, String> tags) {
```

Comment:
tags are purely internal right? Any concern about this blowing up the number of keys?